### DISCUSSION TITLE
v3.0.0

### DISCUSSION THREAD

**Discussion by joschi** [dropwizard/dropwizard#6930] [Announcements]
# Dropwizard 3.0.0

It's finally here, the next major release of Dropwizard! This one has been developed and released in tandem with [Dropwizard 4.0.0](https://github.com/dropwizard/dropwizard/releases/tag/v4.0.0).

If you're wondering what the differences are: Not much!

3️⃣ Dropwizard 3.0.0 is still based on and is using dependencies for Java EE and the `javax.*` package namespace. The effort to migrate from Dropwizard 2.x to Dropwizard 3.0.0 should be minimal for many projects.

4️⃣ Dropwizard 4.0.0 is based on and is using dependencies for Jakarta EE and the `jakarta.*` package namespace. The effort to migrate from Dropwizard 2.x to Dropwizard 4.0.0 might be bigger due to more package changes and more breaking changes in our dependencies.

📣 In any case, let us know how the migration went and if you were missing anything important in the upgrade notes in our GitHub Discussions!

## ⚠️  Breaking changes

For a full list of relevant changes please refer to the [Upgrade Notes for Dropwizard 3.0.x](https://www.dropwizard.io/en/release-3.0.x/manual/upgrade-notes/upgrade-notes-3_0_x.html).

### Java 11 baseline

Dropwizard 3.0.0 and later require Java 11 as a baseline. Later versions of Java are also working.

### Dropwizard Package Structure and JPMS

In order to properly support the Java Platform Module System (JPMS), the Java packages in modules must not overlap, or put differently, the packages may not be split into multiple modules.

Dropwizard 3.0.0 won’t enable full support for the JPMS. Instead, as a transition step, automatic modules are introduced.

Affected packages:

Maven module | Old package | New package
-- | -- | --
`dropwizard-core` | `io.dropwizard` | `io.drowizard.core`
`dropwizard-logging` | `io.dropwizard.logging` | `io.dropwizard.logging.common`
`dropwizard-metrics` | `io.dropwizard.metrics` | `io.dropwizard.metrics.common`
`dropwizard-views` | `io.dropwizard.views` | `io.dropwizard.views.common`

### Jetty 10

The main change introduced in Dropwizard 3.0.0 is the upgrade to Jetty 10.0.x. Jetty 10.0.x is built for Java 11, therefore the Java version change was necessary.

This change comes with some migration cost. For detailed information regarding the changes introduced in Jetty 10.0.x, refer to the [Jetty migration guide](https://www.eclipse.org/jetty/documentation/jetty-10/programming-guide/index.html#pg-migration-94-to-10).

### Apache HttpClient 5

The version of Apache HttpClient used in [dropwizard-client](https://www.dropwizard.io/en/release-3.0.x/manual/client.html) has been updated to version 5.x.

For more information refer to the [Apache HttpClient 5.0 migration guide](https://hc.apache.org/httpcomponents-client-5.1.x/migration-guide/migration-to-classic.html).

### Removed Support for JUnit 4.x

Support for testing with JUnit 4.x has been moved from `dropwizard-testing` to [dropwizard-testing-junit4](https://github.com/dropwizard/dropwizard-testing-junit4).

## 🥇 Special thanks

The release of Dropwizard 3.0.0 wouldn't have been possible without the help of the community (YOU!) and especially the following contributors:

@zUniQueX, @rhowe, @pstackle, @dennyac, and countless others!

## What's Changed since Dropwizard 3.0.0-beta.5
* Update dependency org.apache.tomcat:tomcat-jdbc to v10.1.6 by @renovate in https://github.com/dropwizard/dropwizard/pull/6740
* Update github/codeql-action digest to 32dc499 by @renovate in https://github.com/dropwizard/dropwizard/pull/6739
* Update dependency org.apache.maven.plugins:maven-compiler-plugin to v3.11.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6747
* Update dependency org.apache.maven.plugins:maven-compiler-plugin to v3.11.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6746
* Update dependency org.eclipse.jetty:jetty-bom to v10.0.14 by @renovate in https://github.com/dropwizard/dropwizard/pull/6753
* Update scacap/action-surefire-report digest to a2d14ba by @renovate in https://github.com/dropwizard/dropwizard/pull/6752
* Update dependency org.checkerframework:checker-qual to v3.32.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6763
* Properly declare plugin versions in java-simple archetype by @joschi in https://github.com/dropwizard/dropwizard/pull/6768
* Update dependency org.apache.tomcat:tomcat-jdbc to v10.1.7 by @renovate in https://github.com/dropwizard/dropwizard/pull/6770
* Update dependency com.github.ben-manes.caffeine:caffeine to v3.1.5 by @renovate in https://github.com/dropwizard/dropwizard/pull/6772
* Update dependency net.bytebuddy:byte-buddy to v1.14.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6775
* Update dependency com.uber.nullaway:nullaway to v0.10.10 by @renovate in https://github.com/dropwizard/dropwizard/pull/6781
* Update scacap/action-surefire-report digest to 01b2782 by @renovate in https://github.com/dropwizard/dropwizard/pull/6788
* Revert "Update scacap/action-surefire-report digest to 01b2782" by @joschi in https://github.com/dropwizard/dropwizard/pull/6789
* Update mockito monorepo to v5.2.0 (minor) by @renovate in https://github.com/dropwizard/dropwizard/pull/6791
* Update actions/cache action to v3.3.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6790
* Update github/codeql-action digest to 16964e9 by @renovate in https://github.com/dropwizard/dropwizard/pull/6801
* Update actions/cache action to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6802
* Update dependency maven-wrapper to v3.2.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6803
* Update scacap/action-surefire-report digest to 4845216 by @renovate in https://github.com/dropwizard/dropwizard/pull/6809
* Update dependency maven to v3.8.8 by @renovate in https://github.com/dropwizard/dropwizard/pull/6813
* Update dependency net.bytebuddy:byte-buddy to v1.14.2 by @renovate in https://github.com/dropwizard/dropwizard/pull/6814
* Update dependency org.apache.maven.plugins:maven-surefire-plugin to v3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6815
* Update dependency org.apache.maven.plugins:maven-surefire-plugin to v3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6816
* Update github/codeql-action digest to 168b99b by @renovate in https://github.com/dropwizard/dropwizard/pull/6830
* Update actions/checkout digest to 24cb908 by @renovate in https://github.com/dropwizard/dropwizard/pull/6829
* Update logback.version to v1.3.6 (patch) by @renovate in https://github.com/dropwizard/dropwizard/pull/6832
* Update actions/checkout action to v3.4.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6833
* Update slf4j.version to v2.0.7 (patch) by @renovate in https://github.com/dropwizard/dropwizard/pull/6838
* Update dependency io.dropwizard.metrics:metrics-bom to v4.2.18 by @renovate in https://github.com/dropwizard/dropwizard/pull/6841
* Update dependency maven to v3.9.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6844
* Update dependency org.apache.maven.plugins:maven-help-plugin to v3.4.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6845
* Update dependency org.apache.maven.plugins:maven-release-plugin to v3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6851
* Update dependency com.mattbertolini:liquibase-slf4j to v5 by @renovate in https://github.com/dropwizard/dropwizard/pull/6853
* Update github/codeql-action digest to 67a35a0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6858
* Update actions/stale action to v8 by @renovate in https://github.com/dropwizard/dropwizard/pull/6863
* Use Mockito BOM in dropwizard-dependencies by @joschi in https://github.com/dropwizard/dropwizard/pull/6868
* Update dependency org.glassfish.jersey:jersey-bom to v2.39.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6884
* Update actions/checkout digest to 8f4b7f8 by @renovate in https://github.com/dropwizard/dropwizard/pull/6885
* Update dependency org.apache.maven.plugins:maven-deploy-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6886
* Update dependency org.apache.maven.plugins:maven-deploy-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6887
* Update dependency org.apache.maven.plugins:maven-install-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6888
* Update dependency org.apache.maven.plugins:maven-install-plugin to v3.1.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6889
* Update dependency org.apache.maven.plugins:maven-resources-plugin to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6890
* Update dependency org.apache.maven.plugins:maven-resources-plugin to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6891
* Update dependency org.apache.maven.shared:maven-filtering to v3.3.1 by @renovate in https://github.com/dropwizard/dropwizard/pull/6892
* Update actions/checkout action to v3.5.0 by @renovate in https://github.com/dropwizard/dropwizard/pull/6893
* Set application state healthy if there are no unhealthy critical checks by @joschi in https://github.com/dropwizard/dropwizard/pull/6909
* Update actions/setup-java digest to 5ffc13f by @renovate in https://github.com/dropwizard/dropwizard/pull/6912
* Update github/codeql-action digest to 04df126 by @renovate in https://github.com/dropwizard/dropwizard/pull/6913
* Update dependency net.bytebuddy:byte-buddy to v1.14.3 by @renovate in https://github.com/dropwizard/dropwizard/pull/6918

**Full Changelog**: https://github.com/dropwizard/dropwizard/compare/v3.0.0-beta.5...v3.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/dropwizard/dropwizard/releases/tag/v3.0.0'>v3.0.0</a>.</em>

**Comment by cw-dimedis** [comment]
Hey congrats on the release of v3 and v4. Do you already have an EOL for v2 in mind? We have quite a few projects to upgrade... ;)

**Reply by joschi** [reply]
We were thinking about sunsetting Dropwizard 2.1.x 6 months after the release of Dropwizard 3.0.0 & 4.0.0, so this would be September 30, 2023.

**Reply by cw-dimedis** [reply]
So be it. Thanks, we are going to plan accordingly.
